file.go

```diff
@@ -1,20 +1,132 @@
 package s3fs
 
 import (
+	"errors"
+	"fmt"
 	"io"
 	"io/fs"
+	"io/ioutil"
 	"path"
 	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
 )
 
 var (
 	_ fs.File     = (*file)(nil)
 	_ fs.FileInfo = (*fileInfo)(nil)
+	_ io.Seeker   = (*file)(nil)
 )
 
 type file struct {
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
 	io.ReadCloser
-	stat func() (fs.FileInfo, error)
+	stat   func() (fs.FileInfo, error)
+	offset int64
+}
+
+func openFile(cl s3iface.S3API, bucket string, name string) (fs.File, error) {
+	out, err := cl.GetObject(&s3.GetObjectInput{
+		Key:    &name,
+		Bucket: &bucket,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	statFunc := func() (fs.FileInfo, error) {
+		return stat(cl, bucket, name)
+	}
+
+	if out.ContentLength != nil && out.LastModified != nil {
+		// if we got all the information from GetObjectOutput
+		// then we can cache fileinfo instead of making
+		// another call in case Stat is called.
+		statFunc = func() (fs.FileInfo, error) {
+			return &fileInfo{
+				name:    path.Base(name),
+				size:    *out.ContentLength,
+				modTime: *out.LastModified,
+			}, nil
+		}
+	}
+
+	return &file{
+		cl:         cl,
+		bucket:     bucket,
+		name:       name,
+		ReadCloser: out.Body,
+		stat:       statFunc,
+		offset:     0,
+	}, nil
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.ReadCloser.Read(p)
+	f.offset += int64(n)
+	return n, err
+}
+
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	newOffset := f.offset
+
+	stat, err := f.Stat()
+	if err != nil {
+		return 0, err
+	}
+	size := stat.Size()
+
+	switch whence {
+	case io.SeekStart:
+		newOffset = offset
+	case io.SeekCurrent:
+		newOffset += offset
+	case io.SeekEnd:
+		newOffset = size + offset
+	default:
+		return 0, errors.New("s3fs.file.Seek: invalid whence")
+	}
+
+	// If the position has not moved, there is no need to make a new query
+	if f.offset == newOffset {
+		return newOffset, nil
+	}
+
+	if newOffset < 0 {
+		return 0, errors.New("s3fs.file.Seek: seeked to a negative position")
+	}
+	if newOffset >= size {
+		f.ReadCloser = ioutil.NopCloser(eofReader{})
+		f.offset = newOffset
+		return f.offset, nil
+	}
+
+	err = f.Close()
+	if err != nil {
+		return f.offset, err
+	}
+
+	rawObject, err := f.cl.GetObject(
```

Comment:
There is one more potential big issue here.

With your current code, cached file info (stat func) may no longer be valid. For example, if in the meantime your file has been overwritten and its size changed, your file's size is no longer valid. 

This is fine with os.File, because it points to the same inode, but that's not the case here.

How would you solve this?

Reply:
The current solution is to reassign f.stat (line 135). However, the newly created `getStatFunc` is not very nice, and I would be happy about suggestions how to change it; one reason why it isn't very nice is the fact that I have to pass the current offset to properly calculate the file size (as s3ObjOutput.ContentLength is only the remaining file size).

Reply:
Right, but stat in itself is not the real problem. It's the fact that your file could have been overwritten and is now different

Reply:
I wrote this little app to illustrate the problem: https://gist.github.com/jszwec/f314aa464d1643c55bb84c689368aee8

If you want to easily start it, use one of the following docker compose files before running it:
https://github.com/jszwec/s3fs/blob/main/test/localstack/docker-compose.yml
https://github.com/jszwec/s3fs/blob/main/test/minio/docker-compose.yml

Imagine what would happen if you used this implementation of seeker with your downloader and the file would change in the meantime...

This is a potential blocker for this PR, because correctness is more important than speed. At the very least we may make the seeker implementation optional, otherwise we may break existing code.

Reply:
`NewSeekable` sounds like Java, I don't like it. Instead let's use the `Option` design

```go
type Option func(*S3FS)

func WithReadSeeker(fsys *S3FS) { fsys.readSeeker = true }

func New(cl s3iface.S3API, bucket string, opts ...Option) *S3FS {
	fsys := &S3FS{
		cl:     cl,
		bucket: bucket,
	}
	
	for _, opt := range opts {
	  opt(fsys)
	}
	return fsys
}

type S3FS struct {
	cl     s3iface.S3API
	bucket string
	readSeeker bool
}
```

Reply:
Changed & added the BUG comment.

Was moving the struct below the `New` function intentional?

Reply:
> Was moving the struct below the New function intentional?
no, but either works